File: Javascript/ProjetoBanco/src/contabanco.ts
export abstract class ContaBanco {

    private agencia: string
    private conta: string
    private titular: string
    protected saldo: number

    constructor(
        agencia: string,
        conta: string,
        titular: string,
        saldo: number
    ){
        this.agencia = agencia
        this.conta = conta
        this.titular = titular
        this.saldo = saldo
    }

    /*

    Kotlin - fun identificador(parametros): tipo{}

    Typescript - identificador(parametros): tipo{}

    */

    /*
    Encapsulamento

    public - pode ser acessado de qualquer arquivo
    private - pode ser acessado apenas no arquivo
    protected - pode ser acessado na classe pai e nas filhas

    */

    abstract sacar(valor: number): void

    abstract depositar(valor: number): void

    consultarSaldo(){
        console.log(`Saldo da conta R$${this.saldo}`)
    }

}

File: Kotlin/KotlinPoo/src/main/kotlin/ContaBanco.kt
/*
    Classe Conta Banco
        Atributos - Nome, Agencia, Conta, Saldo
        Métodos - Consultar Saldo, Sacar, Depositar
*/

/*
Construtores são informações que serão exigidas a partir do monento da criação de um
objeto
    Os construtores já serão declarados como atributos da classe
 */

class ContaBanco (
    var titular: String,
    var agencia: String,
    val conta: String,
    var saldo: Double
        ) {

    //Construtores secundários

    var limite = false
    var saldoLimite = 0.0
    var chequeEspecial = false

    constructor(
        titular: String,
        agencia: String,
        conta: String,
        saldo: Double,
        limite: Boolean,
        saldoLimite: Double
    ): this(titular, agencia, conta, saldo){
        this.limite = limite
        this.saldoLimite = saldoLimite
    }

    constructor(
        titular: String,
        agencia: String,
        conta: String,
        saldo: Double,
        limite: Boolean,
        saldoLimite: Double,
        chequeEspecial:  Boolean
    ): this(titular, agencia, conta, saldo, limite, saldoLimite){
        this.chequeEspecial = chequeEspecial
    }

    //Declarando os métodos

    fun consultarSaldo(){
        println("O saldo da conta é R$$saldo")
    }

    fun saque(valor: Double){
        if(saldo > 0.0 && valor <= saldo){
            saldo -= valor
            println("Saque de R$$valor realizado com sucesso!")
        }else{
            println("Saque de R$$valor impossível de ser realizado!")
        }
    }

    fun deposito(valor: Double){
        if(valor > 0.0){
            saldo += valor
            println("Deposito de R$$valor realizado com sucesso!")
        }else{
            println("Deposito de R$$valor impossível de ser realizado!")
        }
    }

}